### PR TITLE
feat(submission): migrate release/review/submit/open from gdcapi

### PR DESCRIPTION
**Pull request by kulgan**

### CODE REVIEW

File: sheepdog/api.py [hunk 1]
```diff
@@ -31,6 +31,10 @@ def app_register_blueprints(app):
     app.register_blueprint(sheepdog_blueprint, url_prefix=v0 + '/submission')
     app.register_blueprint(sheepdog_blueprint, url_prefix='/submission')
 
+    # register legacy
+    app.register_blueprint(sheepdog_blueprint, url_prefix='/legacy/submission')
```
Comment (anmaxvl) [inline sheepdog/api.py:35]: lets not do this. legacy shouldn't support submission at all imo

**Comment by philtom-ctds** [comment]
Amend your commit message to include a jira ticket number.

File: sheepdog/transactions/release/transaction.py [hunk 2]
```diff
@@ -27,24 +28,20 @@ def write_transaction_log(self):
...
-        if project.released:
-            return self.record_error('Project is already released.')
+        project.release_requested = True
 
-        if project.releasable is not True:
-            message = 'Project is not releasable. '\
-                      'Project must be submitted at least once first.'
-            return self.record_error(message)
-
-        project.released = True
```
Comment (philtom-ctds) [inline sheepdog/transactions/release/transaction.py:49]: Looks like this changes from "do the release" to "request the release".   Where does the "do the release" logic happen now?

File: sheepdog/transactions/review/transaction.py [hunk 3]
```diff
@@ -75,6 +82,24 @@ class OpenTransaction(ReviewTransactionBase):
...
+    }
+
     @property
     def to_state(self):
-        return 'open'
+        return False
+
+    def take_action(self):
+        """
+        Mark project `in_rewiew = False` and `request_submission = False`
```
Comment (philtom-ctds) [inline sheepdog/transactions/review/transaction.py:96]:
```suggestion
        Mark project `in_review = False` and `request_submission = False`
```

File: sheepdog/transactions/review/transaction.py [hunk 4]
```diff
@@ -75,6 +82,24 @@ class OpenTransaction(ReviewTransactionBase):
...
+
+        :returns: None
+
+        """
+
+        project = utils.lookup_project(self.db_driver, self.program, self.project)
+        project.in_review = self.to_state
+        project.request_submission = self.to_state
```
Comment (philtom-ctds) [inline sheepdog/transactions/review/transaction.py:104]: Another comment for the future, not now:

It almost looks like `to_state` should return a `dict` that has the keys and associate values.  Then, we can set the appropriate values on `project` without the key being here and the value being in to_state.  It'd also allow us to have different values for different keys.

File: tests/integration/submission/test_async.py [hunk 5]
```diff
@@ -414,7 +412,7 @@ def assert_nodes_in_state(expected_state):
             assert case.state == expected_state
             assert sample.state == expected_state
 
-    assert_nodes_in_state('submitted')
+    assert_nodes_in_state('validated')
```
Comment (philtom-ctds) [inline tests/integration/submission/test_async.py:415]: Could you explain why this changed?  I didn't see a change in the PR that would explain it.  

I'm mostly concerned about changing behavior and what the downstream implications might be.

File: sheepdog/config.py [hunk 6]
```diff
@@ -17,11 +17,19 @@ def create_factory(host, port=443, timeout=10) -> http_client.HTTPSConnection:
 
 
 def https_connection_factory():
-    if os.getenv("S3_ENABLE_HTTPS_CONNECTION_FACTORY", "False").lower() == "true":
+    if getenv_as_bool("S3_ENABLE_HTTPS_CONNECTION_FACTORY", False) is True:
```
[the remark this answers is not in the document]
Comment (kulgan) [reply sheepdog/config.py:20]: This is intentional, more explicit and shows intent.

File: sheepdog/transactions/release/transaction.py (hunk 2 above)
Comment (kulgan) [reply sheepdog/transactions/release/transaction.py:49]: happens in batch_release

File: tests/integration/submission/test_async.py (hunk 5 above)
Comment (kulgan) [reply tests/integration/submission/test_async.py:415]: similar to the comment above, state transition happens in `batch_submit`. Sheepdog is more like a generic submission service which we extended in the gdcapi to be more specific. The workflows implemented by default does not exactly match what we have in gdcapi. In normal sheepdog state transition to release and submitted happens pretty easily but in gdcapi these are done in batches. There has been no change in behavior, what this PR does is to port whatever we have in gdcapi over here

File: sheepdog/transactions/release/transaction.py (hunk 2 above)
Comment (philtom-ctds) [reply sheepdog/transactions/release/transaction.py:49]: ah.  is this the part of the admin api that depends on sheepdog?

File: tests/integration/submission/test_async.py (hunk 5 above)
Comment (philtom-ctds) [reply tests/integration/submission/test_async.py:415]: thanks for the explanation

**Comment by kulgan** [comment]
> The lack of unit tests bothers me. But that's out-of-scope for this PR since it's moving existing code out of gdcapi.
> 
> Most of the classes in this PR would be difficult to unit test. At some point in the future, we should think about refactoring these classes to make unit testing (without mocking) easy.

Sheepdog is design patterns is pretty unique, modules are tightly coupled and hard to tests, but we do have some tests coverage which you can look at in codacy

**Pull request merged**